core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java

```diff
@@ -0,0 +1,262 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch;
+
+import org.elasticsearch.action.RoutingMissingException;
+import org.elasticsearch.action.support.broadcast.BroadcastShardOperationFailedException;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.discovery.DiscoverySettings;
+import org.elasticsearch.index.shard.IndexShardRecoveringException;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.test.ESTestCase;
+import org.hamcrest.Matcher;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import static java.util.Collections.singleton;
+import static org.hamcrest.CoreMatchers.allOf;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.hasItem;
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.Matchers.hasSize;
+
+public class ElasticsearchExceptionTests extends ESTestCase {
+
+    public void testToXContent() throws IOException {
+        ElasticsearchException e = new ElasticsearchException("test");
+        assertExceptionAsJson(e, false, equalTo("{\"type\":\"exception\",\"reason\":\"test\"}"));
+
+        e = new IndexShardRecoveringException(new ShardId("_test", "_0", 5));
+        assertExceptionAsJson(e, false, equalTo("{\"type\":\"index_shard_recovering_exception\"," +
+                "\"reason\":\"CurrentState[RECOVERING] Already recovering\",\"index_uuid\":\"_0\",\"shard\":\"5\",\"index\":\"_test\"}"));
+
+        e = new BroadcastShardOperationFailedException(new ShardId("_index", "_uuid", 12), "foo", new IllegalStateException("bar"));
+        assertExceptionAsJson(e, false, equalTo("{\"type\":\"illegal_state_exception\",\"reason\":\"bar\"}"));
+
+        e = new ElasticsearchException(new IllegalArgumentException("foo"));
+        assertExceptionAsJson(e, false, equalTo("{\"type\":\"exception\",\"reason\":\"java.lang.IllegalArgumentException: foo\"," +
+                "\"caused_by\":{\"type\":\"illegal_argument_exception\",\"reason\":\"foo\"}}"));
+
+        e = new ElasticsearchException("foo", new IllegalStateException("bar"));
+        assertExceptionAsJson(e, false, equalTo("{\"type\":\"exception\",\"reason\":\"foo\"," +
+                "\"caused_by\":{\"type\":\"illegal_state_exception\",\"reason\":\"bar\"}}"));
+        assertExceptionAsJson(e, true, allOf(startsWith("{\"type\":\"exception\",\"reason\":\"foo\""),
```

Comment:
The exception's stack trace is added to the XContent when a special parameter is set, otherwise it is skipped by default. On the REST API the parameter is named `error_trace`, it is binded to the `toXContent.Params` named `rest.exception.stacktrace.skip`.

I added a comment and renamed `printStackTrace` to `errorTrace` in this test. Also, since the stack trace can be large it only checks the beginning fo the JSON: `allOf()` is removed in favor of a single `startsWith` matcher. I hope this is fine now.